File: C_Plus_Plus/p1679.cpp
#include <vector>
#include <algorithm>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int maxOperations(vector<int>& nums, int k) {
        int result = 0;
        sort(nums.begin(), nums.end());
        int left = 0, right = nums.size()-1;
        while(left < right){
            if(nums[left] + nums[right] == k){
                left++;
                right--;
                result++;
            } else if(nums[left] + nums[right] < k){
                left++;
            } else {
                right--;
            }
        }
        return result;
    }
};



    //     int maxOperations(vector<int>& nums, int k) {
    //     int result = 0;
    //     unordered_map<int,int>mp;
    //     for(int i = 0; i < nums.size(); i++){
    //         if(mp[k-nums[i]] > 0){
    //             result++;
    //             mp[k-nums[i]]--;
    //         } else {
    //             mp[nums[i]]++;
    //         }
    //     }
    //     return result;
    // }

File: C_Sharp/1001-2000/p1679.cs
using System.Collections.Immutable;

namespace p1679 {
    public class Solution {
        public int MaxOperations(int[] nums, int k) {
            int result = 0;
            Dictionary<int,int> mp = new Dictionary<int, int>();
            foreach(int num in nums){
                int count;
                if(mp.TryGetValue(k-num, out count) && count > 0){
                    mp[k-num]--;
                    result++;
                } else {
                    mp.TryGetValue(num, out count);
                    mp[num] = count + 1;
                }
            }
            return result;
        }
    }    
}
        // Used immutable array to speed up array scan
        // public int MaxOperations(int[] nums, int k) {
        //     Array.Sort(nums);
        //     int result = 0;
        //     ImmutableArray<int> nums2 = nums.ToImmutableArray();
        //     int left = 0, right = nums.Length - 1;
        //     while(left < right){
        //         if(nums2[left] + nums2[right] == k){
        //             left++;
        //             right--;
        //             result++;
        //         } else if(nums2[left] + nums2[right] < k){
        //             left++;
        //         } else {
        //             right--;
        //         }
        //     }
        //     return result;
        // }